Dynamic Macros

algernon opened the pull request:
This - when complete - will address #370. Currently it is a work in progress.

What this does, is introduce a new plugin: `SequenceMacros` (working title, better suggestions welcome!), which are macros stored in EEPROM, and can be updated via `Focus`. It supports all the macro steps the `Macros` plugin does, but unlike that plugin, this only supports those, and will trigger the macro play immediately on keypress. There's no support for holding or release, since we'd need custom code for that. As such, we only support one active macro at a time, too.

We still need documentation and examples. For now, you'll need to place an `SM(n)` keycode on your keymap somewhere, where `n` is the SequenceMacro index. Then, you need to set a map via Focus, something like this: `focus macros.map 8 4 8 5 0 8 6 8 7 0 0`. This will tap `ab` for `SM(0)` and `cd` for `SM(1)`. `focus macros.map` without arguments will dump the _whole_ reserved area, not just the part occupied by macros, so it may contain garbage in the end. Interpreting the dump is the responsibility of the caller.

The plugin also provides a `macros.trigger N` command, to trigger a macro without having to place it on the keymap.

obra commented:
Just to make sure I understand correctly, it supports steps like the following?

* delay of x ms
* press key down
* release key
* tap key (press and release)

algernon commented:
Yes, it supports all of those.

obra commented:
I agree that it wants a different name. Other than that, it looks ok to me.

Dynamic Macro? Simple Macro?
ᐧ

On Sat, Sep 14, 2019 at 4:48 PM Gergely Nagy <notifications@github.com>
wrote:

> Yes, it supports all of those.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/687?email_source=notifications&email_token=AAALC2HGLARGLLRYJMFQYD3QJVZ5FA5CNFSM4IWXO6I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6XGFAA#issuecomment-531522176>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAALC2F7BZXQUG7IVWAKBALQJVZ5FANCNFSM4IWXO6IQ>
> .
>

algernon commented:
I'm not a fan of "dynamic macro", because to me, that means it can be changed with the keyboard alone (a'la `EEPROM-Keymap-Programmer`). Simple... well, using these is anything but, at least without Chrysalis.

Mind you, almost anything is better than "Sequence Macros", so if we can't find anything better, I'm ok with "dynamic macros" too.

obra commented:
To me, it just means that it can be changed. And the limitation here is not that it can’t be done from the keyboard so much that we don’t have code to do it yet 

> On Sep 15, 2019, at 4:19 PM, Gergely Nagy <notifications@github.com> wrote:
> 
> ﻿
> I'm not a fan of "dynamic macro", because to me, that means it can be changed with the keyboard alone (a'la EEPROM-Keymap-Programmer). Simple... well, using these is anything but, at least without Chrysalis.
> 
> Mind you, almost anything is better than "Sequence Macros", so if we can't find anything better, I'm ok with "dynamic macros" too.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.

algernon commented:
Renamed, and wrote an example sketch. TODO: Documentation, astyle, NEWS.

algernon commented:
After renaming this to dynamic macros, I now added an example, made it astyle-clean, and wrote a bit of documentation too. The documentation isn't very verbose, mind you.

I think this is now ready for review.

algernon commented:
> looks good to me. Will test now on the Raise

The Raise will need a commit call in the Focus hook, FWIW.

obra commented on doc/plugin/DynamicMacros.md:
```diff
@@ -0,0 +1,96 @@
+
+Dynamic macros are similar to [Macros][plugin:macros], but unlike them, they can
+be re-defined without compiling and flashing new firmware: one can change
+dynamic macros via [Focus][plugin:focus], using a tool like
```
changing them via a focus plugin isn't the -only- way one can change them. Do we expose an internal API for changing them, too? So that someone could write an on-keyboard macro writing plugin?

algernon replied:
We do not currently expose an API to change them, the storage pointer is private, and so is the map we build on boot & change. We _could_ expose the storage pointer and size (via an accessor), which would allow a future on-keyboard macro writing plugin to update things.

obra replied:
I do not think we need to add that API to merge this, but I think the doc should break out the fact that there is a Focus API for editing dynamic macros from the fact that we have these neat dynamic macros which are stored in local storage.

obra commented on doc/plugin/DynamicMacros.md:
```diff
@@ -0,0 +1,96 @@
+
+Dynamic macros come with certain limitations, however: unlike the built-in
+macros, dynamic ones do not support running custom code, they can only play back
+a sequence of key events, and do so whenever one presses the dynamic macro key.
```
Is it only -key- events? or do they also support mouse clicks / mousekey events?

algernon replied:
You are correct, the wording here isn't clear: any `Key` is supported. Depending on the step, of course, some only support keys, but the main variants support mouse clicks, mousekeys, and everything else too. Same thing as for the `MACRO()` helper in `Macros`.

obra replied:
fix the docs, then I think this is mergeable

obra commented on doc/plugin/DynamicMacros.md:
```diff
@@ -0,0 +1,96 @@
+macros, dynamic ones do not support running custom code, they can only play back
+a sequence of key events, and do so whenever one presses the dynamic macro key.
+
+You can define up to 32 dynamic macros, there is no limit on their length,
```
where is the 32 hardcoded?

algernon replied:
In [Kaleidoscope-Ranges.h](https://github.com/keyboardio/Kaleidoscope/pull/687/files#diff-767e75f57c447ea243026871575d246bR52-R53). We can easily up the number, mind you.

obra commented on src/Kaleidoscope-Ranges.h:
```diff
@@ -49,6 +49,8 @@ enum : uint16_t {
   REDIAL,
   TURBO,
+  DYNAMIC_MACRO_FIRST,
+  DYNAMIC_MACRO_LAST = DYNAMIC_MACRO_FIRST + 31,
```
I would probably break that magic number out into its own constant defined somewhere else, possibly over in the dynamic macros code? At the very least, just broken out into its definition, just to make it clear that the number is "how many dynamic macros we get"

algernon replied:
Can we do that in a separate PR, where I lift all the other magic numbers in Ranges out? (There's... many of them)

obra replied:
Yeah. That's ok

algernon commented:
@mattvenn discovered a nasty bug in this PR: the focus handler is wrong, it dump and stores 8-bit values only, so there's no way right now to use the 16-bit steps right now.

There are two approaches I can take here:

1. Require the protocol to only send 8-bit values, so any 16 bit values would be split on the wire. This is simple on the firmware side, and not too bad on the Chrysalis side either.
2. Update the Focus handler to be smarter, and interpret macros. This would require a big refactor, because this'd be the third interpreter (one is in `play`, another in `updateDynamicMacroCache`), and that'd be too much.

I'm leaning towards the first method right now, that doesn't require _any_ changes to the code, just a note in the docs (and come to think of it, in the focus handler as well, just to be clear).

obra commented:
Keeping the wire protocol to 8 bit values makes a lot of sense to me. Run
with #1
ᐧ

On Wed, Sep 18, 2019 at 2:21 AM Gergely Nagy <notifications@github.com>
wrote:

> @mattvenn <https://github.com/mattvenn> discovered a nasty bug in this
> PR: the focus handler is wrong, it dump and stores 8-bit values only, so
> there's no way right now to use the 16-bit steps right now.
>
> There are two approaches I can take here:
>
>    1. Require the protocol to only send 8-bit values, so any 16 bit
>    values would be split on the wire. This is simple on the firmware side, and
>    not too bad on the Chrysalis side either.
>    2. Update the Focus handler to be smarter, and interpret macros. This
>    would require a big refactor, because this'd be the third interpreter (one
>    is in play, another in updateDynamicMacroCache), and that'd be too
>    much.
>
> I'm leaning towards the first method right now, that doesn't require *any*
> changes to the code, just a note in the docs (and come to think of it, in
> the focus handler as well, just to be clear).
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/687?email_source=notifications&email_token=AAALC2HUC36SNAKNIFHJSK3QKHXJJA5CNFSM4IWXO6I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD67NKIQ#issuecomment-532600098>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAALC2FEJEYVNYV6ROEUIMTQKHXJJANCNFSM4IWXO6IQ>
> .
>

mattvenn commented:
I've tried this

      while (!::Focus.isEOL()) {
        uint16_t b;
        ::Focus.read(b);

        if(b > 0xFF) {
            EEPROM.update(storage_base_ + pos++, (uint8_t)(b >> 8));
            EEPROM.update(storage_base_ + pos++, (uint8_t)(b & 0x00FF));
        }
        else
            EEPROM.update(storage_base_ + pos++, (uint8_t)(b & 0x00FF));
        EEPROM.commit();

And now I can do 16bit macros

algernon commented:
Updated the documentation: it hopefully makes it clearer that the plugin supports not only *key* events, but all kinds of events, be those mouse keys or other macros, etc. Also added a note to the `macros.map` Focus command docs, noting that it will send and expect a 8-bit memory dump.
